### DISCUSSION TITLE
What's the best practice when you have multiple write interactions on single page / components

### DISCUSSION THREAD

**Discussion by technophile-04** [wevm/wagmi#3786] [Q&A]
We've been exploring the transition from Wagmi v1 to v2 and noticed a shift in the API, particularly regarding write contract interactions. Wagmi v2 introduces a more generalized approach with `useWriteContract`,

In Wagmi v1, we would typically set up a dedicated hook for each contract function interaction, like so:

<details>
<summary>
V1 Example: 
</summary>

```ts
// Its verbose/ definite that the hash will be of setGreetings function transaction hash
const { data: setGreetingsTxHash, writeAsync: writeYourContractGreetings } =  useContractWrite({
  address: YourContract?.address,
  abi: YourContract?.abi,
  functionName: "setGreeting",
  args: ["YourContract Hello world!"],
});

const handleSetGreetings = async (newGreetings: string) => {
  // not wrapping into try / catch, just to keep it concise
  await writeYourContractGreetings({
    args: [newGreetings],
  });
};
```

</details>

However, with Wagmi v2,

<details>
<summary>
V2 Example : 
</summary>

```ts
// Its not verbose/ definite, to which transaction this hash belongs to
const { data: maybeSetGreetingsTxHash, writeContractAsync } = useWriteContract();

const handleSetGreetings = async (newGreetings: string) => {
  // not wrapping into try / catch, just to keep it concise
  await writeContractAsync({
    address: YourContract?.address,
    abi: YourContract?.abi,
    functionName: "setGreeting",
    args: [newGreetings],
  });
};
```

</details>

I'm really curious to hear what the core devs and the community think about dealing with multiple contract functions. Is the preference leaning towards using a single instance of useWriteContract per component / page (the API changes kind of points to this) ? Or is there a trend of creating multiple instances of useWriteContract, each potentially aliased to match the contract function it interacts with, for clarity and tracking?

```ts
const { writeContractAsync: writeSetGreetingsFunc, data: setGreetingsTxHash } = useWriteContract();
const { writeContractAsync: writeSomeAnotherFunc, data: someAnotherFuncHash } = useWriteContract();
```

Multiple instance is kind of making more sense to me since we can keep track of which transaction hash / pending state belongs to which function. But then wouldn't the V1 api made more sense in that case?

Pardon me if I am missing something here, would love to hear other's thoughts, Thanks ! 🙌

**Comment by tmm** [comment]
> Is the preference leaning towards using a single instance of useWriteContract per component

No preference. The API was changed in v2 to not accept arguments via the hook and instead via the result actions (e.g. `writeContract`/`writeContractAsync`) to better align with how TanStack Query mutations work. You can use a single instance if you want (can be helpful to call `reset` if you do this) or use multiple instances (e.g. one hook for each write function).
